Jsx3 andrii

andrii-antiukhov opened the pull request:
In this files i made changes, but not all was refactored to JSX3 (need to do this)

```
client/shared/Fi.re
client/shared/React.re
client/shared/ReactOld.re
```

All components in folder `client/src/components/` was refactored to JSX3. But i did not run compiler, so maybe there are some errors.
@jchavarri Not sure about changing `React.re` to `ReactOld.re` I did this because last time me and Feihong found bug, so i created other `React.re` But i do not know do we need this for JSX3 Sorry. If not need thiod changes, i will remove `ReactOld` from all files in this PR.

jchavarri commented on client/shared/ButterToast.re:
```diff
@@ -69,7 +69,7 @@ module ToastOption = {
   [@bs.deriving abstract]
   type t = {
-    content: contentArgs => React.reactElement,
+    content: contentArgs => ReactOld.reactElement,
```
Using `ReactOld` temporarily is ok while you work on this, but ultimately it's not needed and `ReactOld` should be removed before we merge the jsx3 migration.

andrii-antiukhov replied:
Removed `ReacOld`

jchavarri commented on client/shared/Fi.re:
```diff
@@ -1,248 +1,94 @@
-/* react-icons: feather icon */
-module IconContext = {
```
This was not used?

andrii-antiukhov replied:
Looks like this icon does not used in code.

jchavarri commented on client/src/Router.re:
```diff
@@ -51,7 +51,7 @@ module Unload = {
         None;
       });
-      React.null;
+      ReactOld.null;
```
I don't understand this change :/ why not using `React.null`? One less thing to change later on.

andrii-antiukhov replied:
Remove `ReactOld`

jchavarri commented on client/src/Router.rei:
```diff
@@ -23,7 +23,8 @@ module Unload: {
    */
   module Provider: {
     let make:
-      React.childless => React.component(unit, React.noRetainedProps, unit);
+      ReactOld.childless =>
+      ReactOld.component(unit, ReactOld.noRetainedProps, unit);
```
We should update these signatures to jsx3 as well, e.g.

```reason
[@react.component]
let make: unit => React.element;
```

andrii-antiukhov replied:
Changed.

jchavarri commented on client/src/Router.rei:
```diff
@@ -41,7 +42,11 @@ module Link: {
       ~popup: bool=?,
       ~role: string=?,
-      array(React.reactElement)
+      array(ReactOld.reactElement)
```
We should update this to jsx3 as well:
```reason
[@react.component]
  let make:
    (
       ~route: Route.t,
      ~id: string=?,
      ~title: string=?,
      ~className: string=?,
      ~popup: bool=?,
      ~role: string=?,
      ~children: React.element,
      unit
    ) =>
    React.element;
```

andrii-antiukhov replied:
Changed.

jchavarri commented on client/src/Router.rei:
```diff
@@ -41,7 +42,11 @@ module Link: {
+      array(ReactOld.reactElement)
     ) =>
-    React.component(React.stateless, React.noRetainedProps, React.actionless);
+    ReactOld.component(
+      ReactOld.stateless,
+      ReactOld.noRetainedProps,
+      ReactOld.actionless,
+    );
```
This should be just `React.element`

andrii-antiukhov replied:
Changed.

jchavarri commented on client/src/components/Notify.re:
```diff
@@ -1,6 +1,7 @@
-Modules.require("./Notify.css");
+Modules.import("./Notify.css");
```
Please undo all the changes from `require` to `import`. They will break the app. We want `require`.

andrii-antiukhov replied:
Returned back.

jchavarri commented on client/src/components/Notify.re:
```diff
@@ -1,6 +1,7 @@
 open Utils;
 
+[@react.component]
 let base = (~className, ~sticky=false, message) =>
```
`message` should be a labelled arg. Probably it was meant to be `children` 🤔

andrii-antiukhov replied:
I changed it like labeled. But according code this variable must be string type, not children. Maybe.

jchavarri commented on client/src/components/UI_DateTime.re:
```diff
@@ -19,22 +19,12 @@ module Transformer: {
+let make = (~date: Js.Json.t, ~transformer=Transformer.relative, ~className=?) => {
+  let date = date |> Js.Json.decodeString;
+  switch (date) {
+  | None => ReasonReact.null
```
We can use `React.null` now.

andrii-antiukhov replied:
Replaced `ReasonReact.null` to `React.null`

jchavarri commented on client/src/components/UI_LoadingWrapper.re:
```diff
@@ -8,39 +8,44 @@ type action =
+  let {loading, displayLoading, timeoutId} = state;
+
+  React.useEffect(() => {
+    clearTimeout(timeoutId^);
```
If you look carefully at the previous code, this call to `clearTimeout` was done when unmounting. To keep the same behavior, it should be moved to the returned value of the effect, e.g.

```
 React.useEffect(() => {
    Some(() => clearTimeout(timeoutId^));
}
```

andrii-antiukhov replied:
Thanks! Changed to return some value.

jchavarri commented on client/src/components/UI_LoadingWrapper.re:
```diff
@@ -49,15 +54,13 @@ let make = (~delayMs=1000, ~loading: bool, children) => {
+    [|loading|],
+  );
+
+  displayLoading ? children : ReactOld.null;
```
This is not the same behavior as before. It should be something like `children(displayLoading)`.

andrii-antiukhov replied:
Sorry, i am not sure `children` can be function. But maybe. Thanks!

jchavarri commented on client/src/components/UI_UserInfo.re:
```diff
@@ -1,46 +1,42 @@
+let make = (~userId) => {
+  let query = UserInfoGql.make(~userId, ());
+  <UserInfoComponent variables=query##variables>
+    ...{({result}) =>
```
I don't think spread operator `...` is needed with jsx3. I am pretty sure this won't compile.

andrii-antiukhov replied:
Removed

jchavarri reviewed (commented):
@andrii-antiukhov this is looking good. please take a look at comment, I will merge after they are tackled, thanks!

jchavarri reviewed (approved):
Looks good! Please take a look at these minor things and I think we can merge it ,thanks!

maxlibin commented on client/src/components/UI_DateTime.re:
```diff
@@ -19,22 +19,12 @@ module Transformer: {
-    };
-  },
+[@react.component]
+let make = (~date: Js.Json.t, ~transformer=Transformer.relative, ~className=?) => {
```
This component is already updated in jsx3 branch: https://github.com/Sketch-sh/sketch-sh/pull/321
<img width="299" alt="Screenshot 2022-05-06 at 1 51 57 PM" src="https://user-images.githubusercontent.com/1277038/167074778-eddc58c9-bcbd-4612-80bd-664388ad6658.png">

andrii-antiukhov replied:
@jchavarri @maxlibin Here i added this `[@react.component]` I think it need this code. If no need i will remove.

andrii-antiukhov commented on client/src/components/UI_SketchOwnerInfo.re:
```diff
@@ -1,37 +1,35 @@
-      </Router.Link>
-      {noteLastEdited
+[@react.component]
+let make = (~owner, ~noteLastEdited=?, ~className=?) => {
```
Changed.

andrii-antiukhov commented on client/src/components/UI_FullpageLoading.re:
```diff
@@ -1,4 +1,4 @@
-Modules.import("./UI_FullpageLoading.css");
+Modules.require("./UI_FullpageLoading.css");
```
Changed.

andrii-antiukhov commented:
Hi @jchavarri Today i successfully up sketch on my local Mac and i did some refactoring. So not so bad was today. I refactored `Embed.re` and `ReLoadable.re` and fixed some other issues. Please, review this PR and you can merge it to branch `jsx3` if need.
For me and @maxlibin left to refactor 2 files
```
 src_editor/Editor_Blocks.re
 src_editor/Editor_Links.re
```
But maybe we will do this next time, because this files not so simple to refactor. Thanks!

andrii-antiukhov commented:
@jchavarri Sorry i can not finish properly file `src_editor/Editor_Blocks.re` too complicated for me. I done main part but i do not know how to refactor some old code. So you can look and maybe can fix rest part of code. Sorry.
File `src_editor/Editor_Links.re` was refactored full. Sorry, i do not have types because can not up compiler, difficult to understand logic. So i hope i help you how i can. I tried do everything in best way, but not enough knowledge.
So me and Max waiting for your instructions for the next opensource day. If it is still make sense for us. Thanks!
